_posts/2024-11-19-datafusion-python-udf-comparisons.md

```diff
@@ -0,0 +1,623 @@
+
+As I’m testing, I’m primarily limiting my work to the
+[datafusion-python](https://datafusion.apache.org/python/) project, a wrapper around the Rust
+DataFusion library. This wrapper gives you the speed advantages of keeping all of the data in the
+Rust implementation and the ergonomics of working in Python. Personally, I would prefer to work
+purely in Rust, but I also recognize that since the industry works in Python we should meet the
+people where they are.
+
+## User-Defined Functions
+
+The focus of this post is User-Defined Functions (UDFs). The DataFusion library gives a lot of
+useful functions already for doing DataFrame manipulation. These are going to be similar to those
+you find in other DataFrame libraries. You’ll be able to do simple arithmetic, create substrings of
+columns, or find the average value across a group of rows. These cover most of the use cases
+you’ll need in a DataFrame.
+
+However, there will always arise times when you want a custom function. With UDFs you open a
+world of possibilities in your code. Sometimes there simply isn’t an easy way to use built-in
+functions to achieve your goals.
+
+In the following, I’m going to demonstrate two example use cases. These are based on real world
+problems I’ve encountered. Also I want to demonstrate the approach of “make it work, make it work
+well, make it work fast” that is a motto I’ve seen thrown around in data science.
+
+I will demonstrate three approaches to writing UDFs. In order of increasing performance they are
+
+- Writing a pure Python function to do your computation
+- Using the PyArrow libraries in Python to accelerate your processing
+- Writing a UDF in Rust and exposing it to Python
+
+Additionally I will demonstrate two variants of this. The first will be nearly identical to the
+PyArrow library approach to simplicity of understanding how to connect the Rust code to Python. The
+second version we will do the iteration through the input arrays ourselves to give even greater
+flexibility to the user.
+
+Here are the two example use cases, taken from my own work but generalized.
+
+### Use Case 1: Scalar Function
+
+I have a DataFrame and a list of tuples that I’m interested in. I want to filter out the DataFrame
+to only have values that match those tuples from certain columns in the DataFrame.
+
+To give a concrete example, we will use data generated for the [TPC-H benchmarks](https://www.tpc.org/tpch/).
+Suppose I have a table of sales line items. There are many columns, but I am interested in three: a
+part key (`p_partkey`), supplier key (`p_suppkey`), and return status (`p_returnflag`). I want
+only to return a DataFrame with a specific combination of these three values. That is, I want
+to know if part number 1530 from supplier 4031 was sold (not returned), so I want a specific
+combination of `p_partkey = 1530`, `p_suppkey = 4031`, and `p_returnflag = 'N'`. I have a small
+handful of these combinations I want to return.
```

Comment:
I don't think we should change this blog post / example, but I wanted to point out that you can do this same calculation in SQL

```
> create table foo (x int, y int) as values (1,2), (3,4), (5,6);
0 row(s) fetched.
Elapsed 0.046 seconds.

> select * from foo where (x,y) IN ((3,4), (4,5));
+---+---+
| x | y |
+---+---+
| 3 | 4 |
+---+---+
1 row(s) fetched.
Elapsed 0.018 seconds.

> select * from foo where (x,y) IN ((3,4), (4,5));
+---+---+
| x | y |
+---+---+
| 3 | 4 |
+---+---+
1 row(s) fetched.
Elapsed 0.018 seconds.
```

DataFusion rewrites it internally using `struct`s:

```sql
> select * from foo where {'x': x, 'y': y} IN ({'x': 3, 'y': 4}, {'x': 4, 'y': 5});
+---+---+
| x | y |
+---+---+
| 3 | 4 |
+---+---+
```

```sql
> explain select * from foo where (x,y) IN ((3,4), (4,5));
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                       |
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Filter: CAST(struct(foo.x, foo.y) AS Struct([Field { name: "c0", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c1", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }])) IN ([Struct({c0:3,c1:4}), Struct({c0:4,c1:5})])                                         |
|               |   TableScan: foo projection=[x, y]                                                                                                                                                                                                                                                                                                                         |
| physical_plan | CoalesceBatchesExec: target_batch_size=8192                                                                                                                                                                                                                                                                                                                |
|               |   FilterExec: CAST(struct(x@0, y@1) AS Struct([Field { name: "c0", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c1", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }])) IN ([Literal { value: Struct({c0:3,c1:4}) }, Literal { value: Struct({c0:4,c1:5}) }]) |
|               |     MemoryExec: partitions=1, partition_sizes=[1]                                                                                                                                                                                                                                                                                                          |
|               |                                                                                                                                                                                                                                                                                                                                                            |
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.003 seconds.
```